experimental/federated_plugin/federated_plugin/android/src/main/kotlin/dev/flutter/federated_plugin/FederatedPlugin.kt

```diff
@@ -4,37 +4,122 @@
 import io.flutter.plugin.common.MethodChannel.Result
-import io.flutter.plugin.common.PluginRegistry.Registrar
-
-/** FederatedPlugin */
-public class FederatedPlugin: FlutterPlugin, MethodCallHandler {
-  /// The MethodChannel that will the communication between Flutter and native Android
-  ///
-  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
-  /// when the Flutter Engine is detached from the Activity
-  private lateinit var channel : MethodChannel
-
-  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
-    channel = MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "federated_plugin")
-    channel.setMethodCallHandler(this);
-  }
-
-  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
-    if (call.method == "getPlatformVersion") {
-      result.success("Android ${android.os.Build.VERSION.RELEASE}")
-    } else {
-      result.notImplemented()
-    }
-  }
-
-  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
-    channel.setMethodCallHandler(null)
-  }
+import io.flutter.plugin.common.PluginRegistry
+
+
+class FederatedPlugin : FlutterPlugin, MethodCallHandler, ActivityAware, PluginRegistry.RequestPermissionsResultListener {
+    private lateinit var channel: MethodChannel
+    private lateinit var context: Context
+    private lateinit var activity: Activity
+    private lateinit var result: Result
+    private val REQUEST_CODE = 1001
+
+    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
+        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "location")
+        channel.setMethodCallHandler(this)
+        context = flutterPluginBinding.applicationContext
+    }
+
+    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
+        this.result = result
+        if (call.method == "getLocation") {
+            // Check for the runtime permission if SDK version is greater than 23. If permission
+            // are granted, send the location data back to Dart. If permission are not granted,
+            // request for the runtime permissions.
+            if (android.os.Build.VERSION.SDK_INT > android.os.Build.VERSION_CODES.LOLLIPOP) {
+                if (checkPermissions()) {
+                    fetchLocation()
+                } else {
+                    requestPermissions(activity, arrayOf(
+                            Manifest.permission.ACCESS_FINE_LOCATION,
+                            Manifest.permission.ACCESS_COARSE_LOCATION
+                    ), REQUEST_CODE)
+                }
+            } else {
+                fetchLocation()
+            }
+        } else {
+            result.notImplemented()
+        }
+    }
+
+    // Method to send the last known location of the device to Dart.
+    private fun fetchLocation() {
```

Comment:
This method is doing more than its name implies. Consider refactoring the sending of the result out of this method and back into `onMethodCall`.

Reply:
I guess `onMethodCall` won't be apt. since it means the function is also handling other incoming method calls. Also, `onMethodCall` is already used. I'll try to change it to `provideLocation`. Let me know in the next review if I should change it. Happy to iterate.